internal/store/db.go

```diff
@@ -181,46 +193,206 @@ func (db *DB) checksum(table *TableInfo) (string, error) {
 	switch db.dbType {
 	case DatabaseDriverMysql:
 		tmpl = MySQLChecksumTmpl
 	case DatabaseDriverPostgres:
 		tmpl = PostgresChecksumTmpl
+		// in addition to the template, postgres reuqies the selected schema id
+		// to access objects, it's generally public but it's not guaranteed
 		var schemaName sql.NullString
 		err := db.sqlDB.Get(&schemaName, "SELECT current_schema()")
 		if err != nil {
-			return "", fmt.Errorf("could not get current schema: %w", err)
+			return "", cursorData{}, fmt.Errorf("could not get current schema: %w", err)
 		} else if schemaName.String == "" {
 			q.CurrentSchema = "public"
 		} else {
 			q.CurrentSchema = schemaName.String
 		}
 	default:
-		return "", fmt.Errorf("unrecognized database driver: %s", db.dbType)
+		return "", cursorData{}, fmt.Errorf("unrecognized database driver: %s", db.dbType)
 	}
 
 	t, err := t.Parse(tmpl)
 	if err != nil {
-		return "", fmt.Errorf("could not parse template: %w", err)
+		return "", cursorData{}, fmt.Errorf("could not parse template: %w", err)
+	}
+
+	// pagination query is basically the condtion for the WHERE statement for the
+	// checksum query.
+	paginationQuery, args, err := generateQueryForPagination(db.dbType, table.PrimaryKeys, cursor.cursors)
+	if err != nil {
+		return "", cursorData{}, fmt.Errorf("could not generate query for the cursor: %w", err)
 	}
+	paginationQuery = fmt.Sprintf("%s LIMIT %d", paginationQuery, cursor.limit)
+	q.CursorQuery = paginationQuery
 
 	out := bytes.NewBufferString("")
 	err = t.Execute(out, q)
 	if err != nil {
-		return "", fmt.Errorf("could not execute template: %w", err)
+		return "", cursorData{}, fmt.Errorf("could not execute template: %w", err)
 	}
 
-	var v = []struct {
+	var ints = struct {
 		A uint `db:"a"`
 		B uint `db:"b"`
 		C uint `db:"c"`
 		D uint `db:"d"`
 	}{}
-	err = db.sqlDB.Select(&v, out.String())
+	err = db.sqlDB.Get(&ints, out.String(), args...)
+	if err != nil {
+		return "", cursorData{}, fmt.Errorf("could not select checksum: %w", err)
+	}
+	// we convert the result to a string, it's easier to compare
+	result := fmt.Sprintf("%d%d%d%d", ints.A, ints.B, ints.C, ints.D)
+
+	// the remaining part is about determining the next cursors.
+	// since we don't actually read any rows from the tables for sum operation,
+	// we simply need to pick required rows from the query executed above.
+	var c sq.And
+	for i := range cursor.cursors {
+		c = append(c, sq.Gt{table.PrimaryKeys[i]: cursor.cursors[i]})
+	}
+
+	builder := sq.StatementBuilder.PlaceholderFormat(sq.Question)
+	tableName := table.TableName
+	if db.dbType == DatabaseDriverPostgres {
+		builder = sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+		tableName = strings.Join([]string{q.CurrentSchema, tableName}, ".")
+	}
+
+	cursorQueryBuilder := builder.
+		Select(table.PrimaryKeys...).
+		From(tableName).
+		Where(c).
+		OrderBy(strings.Join(table.PrimaryKeys, ",") + " ASC").
+		Limit(uint64(cursor.limit))
```

Comment:
To correctly paginate, we need to have an offset as well right? I am not seeing that here. :thinking: Otherwise, how can we move to the next page?

Reply:
So the first page will have empty cursors and the limit, but second page will have the offset as the cursors and the limit. So it will iterate that way.